pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -194,76 +195,73 @@ func (exp *flowExporter) sendFlowRecords() error {
+	elements := make([]*ipfixentities.InfoElementWithValue, 0)
 
 	for _, ie := range IANAInfoElements {
 		element, err := exp.registry.GetInfoElement(ie, ipfixregistry.IANAEnterpriseID)
 		if err != nil {
 			return 0, fmt.Errorf("%s not present. returned error: %v", ie, err)
 		}
-		if _, err = templateRec.AddInfoElement(element, nil); err != nil {
-			return 0, fmt.Errorf("error when adding %s to template: %v", element.Name, err)
-		}
+		ieWithValue := ipfixentities.NewInfoElementWithValue(element, nil)
+		elements = append(elements, ieWithValue)
 	}
 	for _, ie := range IANAReverseInfoElements {
 		element, err := exp.registry.GetInfoElement(ie, ipfixregistry.IANAReversedEnterpriseID)
 		if err != nil {
 			return 0, fmt.Errorf("%s not present. returned error: %v", ie, err)
 		}
-		if _, err = templateRec.AddInfoElement(element, nil); err != nil {
-			return 0, fmt.Errorf("error when adding %s to template: %v", element.Name, err)
-		}
+		ieWithValue := ipfixentities.NewInfoElementWithValue(element, nil)
+		elements = append(elements, ieWithValue)
 	}
 	for _, ie := range AntreaInfoElements {
 		element, err := exp.registry.GetInfoElement(ie, ipfixregistry.AntreaEnterpriseID)
 		if err != nil {
 			return 0, fmt.Errorf("information element %s is not present in Antrea registry", ie)
 		}
-		if _, err := templateRec.AddInfoElement(element, nil); err != nil {
-			return 0, fmt.Errorf("error when adding %s to template: %v", element.Name, err)
-		}
+		ieWithValue := ipfixentities.NewInfoElementWithValue(element, nil)
+		elements = append(elements, ieWithValue)
 	}
 
-	sentBytes, err := exp.process.AddRecordAndSendMsg(ipfixentities.Template, templateRec.GetRecord())
+	templateSet.AddRecord(elements, templateID)
+
+	sentBytes, err := exp.process.AddSetAndSendMsg(ipfixentities.Template, templateSet.GetSet())
 	if err != nil {
 		return 0, fmt.Errorf("error in IPFIX exporting process when sending template record: %v", err)
 	}
 
 	// Get all elements from template record.
-	exp.elementsList = templateRec.GetTemplateElements()
+	exp.elementsList = elements
 
 	return sentBytes, nil
 }
 
-func (exp *flowExporter) sendDataRecord(dataRec ipfix.IPFIXRecord, record flowexporter.FlowRecord) error {
+func (exp *flowExporter) sendDataSet(dataSet ipfix.IPFIXSet, record flowexporter.FlowRecord, templateID uint16) error {
 	nodeName, _ := env.GetNodeName()
+	elements := make([]*ipfixentities.InfoElementWithValue, 0)
+
 	// Iterate over all infoElements in the list
 	for _, ie := range exp.elementsList {
-		var err error
-		switch ieName := ie.Name; ieName {
+		var ieWithValue *ipfixentities.InfoElementWithValue
+		switch ieName := ie.Element.Name; ieName {
 		case "flowStartSeconds":
-			_, err = dataRec.AddInfoElement(ie, uint32(record.Conn.StartTime.Unix()))
+			ieWithValue = ipfixentities.NewInfoElementWithValue(ie.Element, uint32(record.Conn.StartTime.Unix()))
```

Comment:
Can't we use the value field in ie? Do we need a new variable ieWithValue? We can skip another allocation of elements in L241.

Reply:
Done. Thanks!